File: packages/shared/src/compare.ts
import { CHARACTER_GUESS_FIELDS, HAIR_COLOR_LABELS } from "./fields";
import type {
  Character,
  FeedbackStatus,
  FieldFeedback,
  GuessField,
  GuessFieldKey,
  GuessResult,
} from "./types";

const statusToSymbol = (status: FeedbackStatus): FieldFeedback["symbol"] => {
  if (status === "exact") return "O";
  if (status === "partial") return "~";
  if (status === "higher") return "↑";
  if (status === "lower") return "↓";
  if (status === "unknown") return "?";
  return "X";
};

const sameSet = (left: string[], right: string[]) => {
  if (left.length !== right.length) return false;
  const rightSet = new Set(right);
  return left.every((item) => rightSet.has(item));
};

const hasIntersection = (left: string[], right: string[]) => {
  const rightSet = new Set(right);
  return left.some((item) => rightSet.has(item));
};

const compareMultiSet = (
  guessValues: string[],
  answerValues: string[],
): FeedbackStatus => {
  if (!guessValues.length || !answerValues.length) return "unknown";
  if (sameSet(guessValues, answerValues)) return "exact";
  if (hasIntersection(guessValues, answerValues)) return "partial";
  return "miss";
};

const displayValuesForField = (
  character: Character,
  field: GuessFieldKey,
): string[] => {
  if (field === "firstAppearance") return [character.firstAppearance.workTitle];
  if (field === "releaseYear")
    return [String(character.firstAppearance.releaseYear)];
  if (field === "hairColors")
    return character.hairColors.map((color) => HAIR_COLOR_LABELS[color]);
  return character[field];
};

const valuesForField = (
  character: Character,
  field: GuessFieldKey,
): string[] => {
  if (field === "firstAppearance") return [character.firstAppearance.workId];
  if (field === "releaseYear")
    return [String(character.firstAppearance.releaseYear)];
  if (field === "hairColors") return character.hairColors;
  return character[field];
};

export const compareField = (
  guess: Character,
  answer: Character,
  field: GuessField,
): FieldFeedback => {
  let status: FeedbackStatus = "unknown";

  if (field.compareStrategy === "firstAppearance") {
    if (guess.firstAppearance.workId === answer.firstAppearance.workId) {
      status = "exact";
    } else if (
      guess.firstAppearance.workType === answer.firstAppearance.workType
    ) {
      status = "partial";
    } else {
      status = "miss";
    }
  }

  if (field.compareStrategy === "numberDirection") {
    const guessYear = guess.firstAppearance.releaseYear;
    const answerYear = answer.firstAppearance.releaseYear;
    if (guessYear === answerYear) status = "exact";
    else status = guessYear < answerYear ? "higher" : "lower";
  }

  if (field.compareStrategy === "multiSet") {
    status = compareMultiSet(
      valuesForField(guess, field.key),
      valuesForField(answer, field.key),
    );
  }

  return {
    field: field.key,
    label: field.label,
    status,
    symbol: statusToSymbol(status),
    displayValue: displayValuesForField(guess, field.key),
  };
};

export const compareCharacter = (
  guess: Character,
  answer: Character,
  fields: GuessField[] = CHARACTER_GUESS_FIELDS,
): GuessResult => ({
  guessId: guess.id,
  guessName: guess.names.zhHans,
  guessAvatarUrl: guess.avatarUrl,
  isCorrect: guess.id === answer.id,
  feedback: fields
    .filter((field) => field.visible)
    .map((field) => compareField(guess, answer, field)),
});

File: apps/api/internal/game/compare.go
package game

// HairColorLabel 对应 shared 的 HAIR_COLOR_LABELS。
var HairColorLabels = map[string]string{
	"black":      "黑",
	"brown":      "棕",
	"blonde":     "金",
	"white":      "白",
	"silver":     "银",
	"red":        "红",
	"pink":       "粉",
	"purple":     "紫",
	"blue":       "蓝",
	"green":      "绿",
	"orange":     "橙",
	"gray":       "灰",
	"multicolor": "多色",
	"other":      "其他",
	"none":       "无",
}

// CHARACTER_GUESS_FIELDS 对应 shared 的 CHARACTER_GUESS_FIELDS。
var CharacterGuessFields = CharacterFields.AllFields()

// GameContentDefinition 对应 shared 的 GAME_CONTENT_DEFINITIONS.character。
var GameContentDefinition = struct {
	Label      string
	MaxGuesses int
	Fields     []GuessField
}{
	Label:      "角色",
	MaxGuesses: QuestionScopeDefaultGuesses,
	Fields:     CharacterGuessFields,
}

func StatusToSymbol(status FeedbackStatus) string {
	switch status {
	case FeedbackExact:
		return "O"
	case FeedbackPartial:
		return "~"
	case FeedbackHigher:
		return "↑"
	case FeedbackLower:
		return "↓"
	case FeedbackUnknown:
		return "?"
	default:
		return "X"
	}
}

// displayValuesForField 返回展示用值（发色为中文标签，年份为字符串）。
func DisplayValuesForField(character Character, field GuessFieldKey) []string {
	return CharacterFields.DisplayValues(character, field)
}

// CompareField 对应 shared 的 compareField。
func CompareField(guess, answer Character, field GuessField) FieldFeedback {
	status := CharacterFields.CompareFeedback(guess, answer, field)

	return FieldFeedback{
		Field:        field.Key,
		Label:        field.Label,
		Status:       status,
		Symbol:       StatusToSymbol(status),
		DisplayValue: DisplayValuesForField(guess, field.Key),
	}
}

func CompareCharacterWithMatch(guess, answer Character, fields []GuessField, match MatchResult) GuessResult {
	if fields == nil {
		fields = CharacterGuessFields
	}
	feedback := make([]FieldFeedback, 0, len(fields))
	for _, field := range fields {
		if field.Visible {
			feedback = append(feedback, CompareField(guess, answer, field))
		}
	}
	return GuessResult{
		Kind:           "guess",
		GuessID:        guess.ID,
		GuessName:      guess.Names.ZhHans,
		GuessAvatarURL: guess.AvatarURL,
		IsCorrect:      match.Correct,
		MatchKind:      match.Kind,
		Feedback:       feedback,
	}
}

// CompareCharacter preserves strict identity semantics while callers migrate to GuessEvaluator.
func CompareCharacter(guess, answer Character, fields []GuessField) GuessResult {
	return CompareCharacterWithMatch(guess, answer, fields, StrictIdentityMatcher{}.Match(nil, answer.ID, guess.ID))
}
